File: src/lowlevel/client/icefsoperators/icefsOpen.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:40:54
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsOpen.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsOpen(fuse_req_t fuseReq, fuse_ino_t inode,
                              struct fuse_file_info *fi) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsOpenReq req;
      icefsgrpc::IcefsOpenRes res;
      grpc::ClientContext ctx;

      req.set_inode(inode);

      if (this->clientConfig.cacheMode && (fi->flags & O_ACCMODE) == O_WRONLY) {
        fi->flags &= ~O_ACCMODE;
        fi->flags |= O_RDWR;
      }

      if (this->clientConfig.cacheMode && fi->flags & O_APPEND) {
        fi->flags &= ~O_APPEND;
      }
      req.set_flags(fi->flags);

      grpc::Status status = gRpcClient->DoIcefsOpen(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fi->fh = res.fh();
        if (this->clientConfig.cacheMode) {
          fi->keep_cache = 1;
        } else if ((fi->flags & O_ACCMODE) == O_RDONLY) {
          fi->noflush = 1;
          fi->direct_io = 1;
        } else {
          fi->direct_io = 1;
        }
        fuse_reply_open(fuseReq, fi);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsOpenReq req;
      icefsthrift::IcefsOpenRes res;

      req.__set_inode(inode);

      if (this->clientConfig.cacheMode && (fi->flags & O_ACCMODE) == O_WRONLY) {
        fi->flags &= ~O_ACCMODE;
        fi->flags |= O_RDWR;
      }

      if (this->clientConfig.cacheMode && fi->flags & O_APPEND) {
        fi->flags &= ~O_APPEND;
      }

      req.__set_flags(fi->flags);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsOpen(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      if (!res.status) {
        fi->fh = res.fh;
        if (this->clientConfig.cacheMode) {
          fi->keep_cache = 1;
        } else if ((fi->flags & O_ACCMODE) == O_RDONLY) {
          fi->noflush = 1;
          fi->direct_io = 1;
        } else {
          fi->direct_io = 1;
        }
        fuse_reply_open(fuseReq, fi);
      } else {
        fuse_reply_err(fuseReq, res.status);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}

File: src/lowlevel/server/icefsoperators/icefsOpen.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 07:58:23
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsOpen.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"fmt"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsOpen(fakeInode uint64, flags int32) (status int32, fh uint64) {
	var procName string
	var fd int
	var err error
	s.inodeCacheLock.RLock()
	inode := s.getIcefsInode(fakeInode)
	if inode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}
	inode.inodeLock.Lock()
	s.inodeCacheLock.RUnlock()
	procName = fmt.Sprintf("/proc/self/fd/%v", inode.fd)
	fd, err = syscall.Open(procName, int(flags&(^syscall.O_NOFOLLOW)), 0)
	if err != nil {
		inode.inodeLock.Unlock()
		status = icefserror.IcefsStdErrno(err)
		return
	}
	inode.nopen++
	inode.inodeLock.Unlock()
	status = icefserror.ICEFS_EOK
	fh = uint64(fd)
	return
}

func (s *IcefsGRpcServer) DoIcefsOpen(ctx context.Context, req *pb.IcefsOpenReq) (*pb.IcefsOpenRes, error) {
	var res pb.IcefsOpenRes

	res.Status, res.Fh = s.server.doIcefsOpen(req.Inode, req.Flags)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsOpen(ctx context.Context, req *icefsthrift.IcefsOpenReq) (*icefsthrift.IcefsOpenRes, error) {
	var res icefsthrift.IcefsOpenRes
	var fh uint64

	res.Status, fh = s.server.doIcefsOpen(uint64(req.Inode), req.Flags)
	res.Fh = icefsthrift.Ui64(fh)

	return &res, nil
}
